pkg/runtime/grpc/instances.go

```diff
@@ -147,17 +147,49 @@ func (r *Runtime) createGadgetInstance(gadgetCtx runtime.GadgetContext, runtimeP
 		EventBufferLength: runtimeParams.Get(ParamEventBufferLength).AsInt32(), // default for now
 	}
 
+	targets, explicitTargets, err := r.getTargets(gadgetCtx.Context(), runtimeParams)
+	if err != nil {
+		return fmt.Errorf("getting targets: %w", err)
+	}
+
+	// if targets have explicitly been listed, add them to the `Nodes` list
+	if explicitTargets {
+		for _, t := range targets {
+			instanceRequest.GadgetInstance.Nodes = append(instanceRequest.GadgetInstance.Nodes, t.node)
+		}
+	}
```

Comment:
What do you think about this alternative:

(validation is done a bit later but it seems to work fine as well. Only difference is that explicit was always true for direct connection, is that relevant?)

```diff
diff --git a/pkg/runtime/grpc/grpc-runtime.go b/pkg/runtime/grpc/grpc-runtime.go
index cc164ea2e..25e429c4c 100644
--- a/pkg/runtime/grpc/grpc-runtime.go
+++ b/pkg/runtime/grpc/grpc-runtime.go
@@ -310,7 +310,7 @@ nodesLoop:
 
 // getTargets returns targets depending on the params given and the environment. The returned
 // bool is true, if the user explicitly selected the nodes using params.
-func (r *Runtime) getTargets(ctx context.Context, params *params.Params) ([]target, bool, error) {
+func (r *Runtime) getTargets(ctx context.Context, params *params.Params) ([]target, error) {
 	switch r.connectionMode {
 	case ConnectionModeKubernetesProxy:
 		// Get nodes to run on
@@ -318,19 +318,19 @@ func (r *Runtime) getTargets(ctx context.Context, params *params.Params) ([]targ
 		gadgetNamespace := r.globalParams.Get(ParamGadgetNamespace).AsString()
 		pods, err := getGadgetPods(ctx, r.restConfig, nodes, gadgetNamespace)
 		if err != nil {
-			return nil, false, fmt.Errorf("get gadget pods: %w", err)
+			return nil, fmt.Errorf("get gadget pods: %w", err)
 		}
 		if len(pods) == 0 {
-			return nil, false, fmt.Errorf("get gadget pods: Inspektor Gadget is not running on the requested node(s): %v", nodes)
+			return nil, fmt.Errorf("get gadget pods: Inspektor Gadget is not running on the requested node(s): %v", nodes)
 		}
-		return pods, len(nodes) != 0, nil
+		return pods, nil
 	case ConnectionModeDirect:
 		inTargets := r.globalParams.Get(ParamRemoteAddress).AsStringSlice()
 		targets := make([]target, 0)
 		for _, t := range inTargets {
 			purl, err := url.Parse(t)
 			if err != nil {
-				return nil, false, fmt.Errorf("invalid remote address %q: %w", t, err)
+				return nil, fmt.Errorf("invalid remote address %q: %w", t, err)
 			}
 			tg := target{
 				addressOrPod: purl.Host,
@@ -343,9 +343,9 @@ func (r *Runtime) getTargets(ctx context.Context, params *params.Params) ([]targ
 			}
 			targets = append(targets, tg)
 		}
-		return targets, true, nil
+		return targets, nil
 	}
-	return nil, false, fmt.Errorf("unsupported connection mode")
+	return nil, fmt.Errorf("unsupported connection mode")
 }
 
 func (r *Runtime) RunBuiltInGadget(gadgetCtx runtime.GadgetContext) (runtime.CombinedGadgetResult, error) {
@@ -362,7 +362,7 @@ func (r *Runtime) RunBuiltInGadget(gadgetCtx runtime.GadgetContext) (runtime.Com
 		gadgetCtx.Logger().Debugf("- %s: %q", k, v)
 	}
 
-	targets, _, err := r.getTargets(gadgetCtx.Context(), gadgetCtx.RuntimeParams())
+	targets, err := r.getTargets(gadgetCtx.Context(), gadgetCtx.RuntimeParams())
 	if err != nil {
 		return nil, fmt.Errorf("getting target nodes: %w", err)
 	}
@@ -370,7 +370,7 @@ func (r *Runtime) RunBuiltInGadget(gadgetCtx runtime.GadgetContext) (runtime.Com
 }
 
 func (r *Runtime) getConnToRandomTarget(ctx context.Context, runtimeParams *params.Params) (*grpc.ClientConn, error) {
-	targets, _, err := r.getTargets(ctx, runtimeParams)
+	targets, err := r.getTargets(ctx, runtimeParams)
 	if err != nil {
 		return nil, err
 	}
diff --git a/pkg/runtime/grpc/instances.go b/pkg/runtime/grpc/instances.go
index 18322b2d1..f8df91f71 100644
--- a/pkg/runtime/grpc/instances.go
+++ b/pkg/runtime/grpc/instances.go
@@ -70,7 +70,7 @@ func (r *Runtime) GetGadgetInstances(ctx context.Context, runtimeParams *params.
 func (r *Runtime) runInstanceManagerClientForTargets(ctx context.Context, runtimeParams *params.Params, fn func(target target, client api.GadgetInstanceManagerClient) error) error {
 	// depending on the environment, we need to either connect to a single random target (k8s, where k8s/etcd handles
 	// synchronizing gadget configuration), or all possible targets (ig-daemon).
-	targets, _, err := r.getTargets(ctx, runtimeParams)
+	targets, err := r.getTargets(ctx, runtimeParams)
 	if err != nil {
 		return fmt.Errorf("getting targets: %w", err)
 	}
@@ -147,16 +147,9 @@ func (r *Runtime) createGadgetInstance(gadgetCtx runtime.GadgetContext, runtimeP
 		EventBufferLength: runtimeParams.Get(ParamEventBufferLength).AsInt32(), // default for now
 	}
 
-	targets, explicitTargets, err := r.getTargets(gadgetCtx.Context(), runtimeParams)
-	if err != nil {
-		return fmt.Errorf("getting targets: %w", err)
-	}
-
 	// if targets have explicitly been listed, add them to the `Nodes` list
-	if explicitTargets {
-		for _, t := range targets {
-			instanceRequest.GadgetInstance.Nodes = append(instanceRequest.GadgetInstance.Nodes, t.node)
-		}
+	if paramNode := runtimeParams.Get(ParamNode); paramNode != nil {
+		instanceRequest.GadgetInstance.Nodes = paramNode.AsStringSlice()
 	}
 
 	var listMutex sync.Mutex
diff --git a/pkg/runtime/grpc/oci.go b/pkg/runtime/grpc/oci.go
index 446a19773..ef23fefd7 100644
--- a/pkg/runtime/grpc/oci.go
+++ b/pkg/runtime/grpc/oci.go
@@ -81,7 +81,7 @@ func (r *Runtime) RunGadget(gadgetCtx runtime.GadgetContext, runtimeParams *para
 		return r.createGadgetInstance(gadgetCtx, runtimeParams, paramValues)
 	}
 
-	targets, _, err := r.getTargets(gadgetCtx.Context(), runtimeParams)
+	targets, err := r.getTargets(gadgetCtx.Context(), runtimeParams)
 	if err != nil {
 		return fmt.Errorf("getting target nodes: %w", err)
 	}

```

Reply:
You're right, this looks simpler. I think it covers all cases as well.